### PR TITLE
Refactor/document type objects

### PR DESCRIPTION
**Pull request by glyn**
This change is incomplete and is for review in draft form only.

### CODE REVIEW

File: pkg/controllers/reconcilers.go [hunk 1]
```diff
@@ -67,8 +87,18 @@ type ParentReconciler struct {
 	Config
 }
 
+// NewParentReconciler returns a ParentReconciler given an empty object to represent the parent's type, a Config,
+// and a list of SubReconcilers.
+func NewParentReconciler(objType runtime.Object, cfg Config, subReconcilers ...SubReconciler) *ParentReconciler {
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: I'm 👎 on constructors that mask the semantic value of field names for positional args without adding meaningful behavior.

Comment (glyn) [reply pkg/controllers/reconcilers.go]: Ok, I've inlined the constructor for now. We can always introduce a constructor or builder later if necessary.

File: pkg/controllers/reconcilers.go [hunk 2]
```diff
@@ -54,14 +54,34 @@ type Config struct {
...
+
 // ParentReconciler is a controller-runtime reconciler that reconciles a given
 // existing resource. The ParentType resource is fetched for the reconciler
 // request and passed in turn to each SubReconciler. Finally, the reconciled
 // resource's status is compared with the original status, updating the API
 // server if needed.
 type ParentReconciler struct {
-	// Type of resource to reconcile
-	Type runtime.Object
+	// ObjectType of resource to reconcile
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: godoc should be use the field name, not the type

File: pkg/controllers/reconcilers.go [hunk 3]
```diff
@@ -54,14 +54,34 @@ type Config struct {
...
+}
+
+// NewObjectType constructs an ObjectType from the given empty object.
+func NewObjectType(obj runtime.Object) ObjectType {
+	return ObjectType{obj}
+}
+
+// asObject returns a deep copy of the object type as an apis.Object.
+func (ot ObjectType) asObject() apis.Object {
+	return ot.DeepCopyObject().(apis.Object)
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: by convention, methods do not return copies. It's the responsibility of the caller to create a copy before inflicting mutations that should be localized.

File: pkg/controllers/reconcilers.go [hunk 4]
```diff
@@ -54,14 +54,34 @@ type Config struct {
...
+	runtime.Object
+}
+
+// NewObjectType constructs an ObjectType from the given empty object.
+func NewObjectType(obj runtime.Object) ObjectType {
+	return ObjectType{obj}
+}
+
+// asObject returns a deep copy of the object type as an apis.Object.
+func (ot ObjectType) asObject() apis.Object {
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: should be exported, or drop the method and lean on the current `Object` field

```suggestion
func (ot ObjectType) Object() apis.Object {
```

File: pkg/controllers/reconcilers.go [hunk 5]
```diff
@@ -54,14 +54,34 @@ type Config struct {
 	Tracker   tracker.Tracker
 }
 
+// ObjectType represents the type of a Kubernetes object. Fields of type ObjectType are
+// initialized with an empty object and not modified subsequently.
+type ObjectType struct {
+	runtime.Object
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: this shouldn't be exported

File: pkg/controllers/reconcilers.go [hunk 6]
```diff
@@ -54,14 +54,34 @@ type Config struct {
...
+}
+
+// NewObjectType constructs an ObjectType from the given empty object.
+func NewObjectType(obj runtime.Object) ObjectType {
+	return ObjectType{obj}
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: use named fields

File: pkg/controllers/reconcilers.go [hunk 7]
```diff
@@ -54,14 +54,34 @@ type Config struct {
...
+}
+
+// asObject returns a deep copy of the object type as an apis.Object.
+func (ot ObjectType) asObject() apis.Object {
+	return ot.DeepCopyObject().(apis.Object)
+}
+
+func (ot ObjectType) typeName() string {
```
Comment (scothis) [inline pkg/controllers/reconcilers.go]: should be exported

```suggestion
func (ot ObjectType) Name() string {
```

**Comment by glyn** [comment]
> Even with my inline suggestions, I think we should abandon this PR:
> 
> 1. controller-runtime uses bare `runtime.Object`s to denote type information. We should strive to use consistent idioms when possible
> 2. ObjectType doesn't simplify our abstract reconcilers
> 3. ObjectType inflicts reconciler implementors with one more thing they need to know (granted it's small)

I'm comfortable with that. Closing.

**Pull request closed without merging**